### ISSUE TITLE
Option --no-write-lock-file to `use flake` is handled inconsistently

### ISSUE THREAD

**Issue by alexshpilkin** [nix-community/nix-direnv#189]
Unlike the stdlib implementation of `use flake`, `nix-direnv` will write a lock file and attempt to add it to Git even if it is invoked with `--no-write-lock-file`. As a bonus, it will make a spurious refresh on the second invocation. As far as I can tell, this happens because, while the [`nix print-dev-env` invocation](https://github.com/nix-community/nix-direnv/blob/6dbc942fea1bd5e353cebdf7851fe0be45e6ec22/direnvrc#L186) does receive the arguments the user passed to `use flake`, the later [`nix flake archive` invocation](https://github.com/nix-community/nix-direnv/blob/6dbc942fea1bd5e353cebdf7851fe0be45e6ec22/direnvrc#L201) does not. I’m not sure if just passing `"$@"` there as well would be fine, so I wrote this report instead.

Reproduction:

```shellsession
$ cd /tmp
$ mkdir lockfile-bug && cd $_
$ git init
Initialized empty Git repository in /tmp/lockfile-bug/.git/
$ nix flake init -t github:nix-community/nix-direnv
wrote: /tmp/lockfile-bug/.envrc
wrote: /tmp/lockfile-bug/flake.nix
direnv: error /tmp/lockfile-bug/.envrc is blocked. Run `direnv allow` to approve its content
$ sed -i 's/use flake$/& . --no-write-lock-file/' .envrc
direnv: error /tmp/lockfile-bug/.envrc is blocked. Run `direnv allow` to approve its content
$ direnv allow
direnv: loading /tmp/lockfile-bug/.envrc
direnv: loading https://raw.githubusercontent.com/nix-community/nix-direnv/2.1.0/direnvrc (sha256-FAT2R9yYvVg516v3LiogjIc8YfsbWbMM/itqWsm5xTA=)
direnv: using flake . --no-write-lock-file
warning: Git tree '/tmp/lockfile-bug' is dirty
warning: not writing modified lock file of flake 'git+file:///tmp/lockfile-bug':
• Added input 'flake-utils':                                                                                                                                   
    'github:numtide/flake-utils/1ed9fb1935d260de5fe1c2f7ee0ebaae17ed2fa1' (2022-05-30)
• Added input 'nixpkgs':
    'github:NixOS/nixpkgs/17b62c338f2a0862a58bb6951556beecd98ccda9' (2022-05-28)
warning: Git tree '/tmp/lockfile-bug' is dirty
warning: creating lock file '/tmp/lockfile-bug/flake.lock'
warning: Git tree '/tmp/lockfile-bug' is dirty
direnv: nix-direnv: renewed cache
direnv: export +AR +AS +CC +CONFIG_SHELL +CXX +HOST_PATH +IN_NIX_SHELL +LD +NIX_BINTOOLS +NIX_BINTOOLS_WRAPPER_TARGET_HOST_x86_64_unknown_linux_gnu +NIX_BUILD_CORES +NIX_CC +NIX_CC_WRAPPER_TARGET_HOST_x86_64_unknown_linux_gnu +NIX_CFLAGS_COMPILE +NIX_ENFORCE_NO_NATIVE +NIX_HARDENING_ENABLE +NIX_INDENT_MAKE +NIX_LDFLAGS +NIX_STORE +NM +OBJCOPY +OBJDUMP +RANLIB +READELF +SIZE +SOURCE_DATE_EPOCH +STRINGS +STRIP +buildInputs +buildPhase +builder +configureFlags +depsBuildBuild +depsBuildBuildPropagated +depsBuildTarget +depsBuildTargetPropagated +depsHostHost +depsHostHostPropagated +depsTargetTarget +depsTargetTargetPropagated +doCheck +doInstallCheck +dontAddDisableDepTrack +name +nativeBuildInputs +out +outputs +patches +phases +propagatedBuildInputs +propagatedNativeBuildInputs +shell
 +shellHook +stdenv +strictDeps +system ~PATH ~XDG_DATA_DIRS
$ git status
On branch master

No commits yet

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        new file:   .envrc
        new file:   flake.lock
        new file:   flake.nix

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        .direnv/

no changes added to commit (use "git add" and/or "git commit -a")
direnv: loading /tmp/lockfile-bug/.envrc
direnv: loading https://raw.githubusercontent.com/nix-community/nix-direnv/2.1.0/direnvrc (sha256-FAT2R9yYvVg516v3LiogjIc8YfsbWbMM/itqWsm5xTA=)
direnv: using flake . --no-write-lock-file
warning: Git tree '/tmp/lockfile-bug' is dirty
warning: Git tree '/tmp/lockfile-bug' is dirty
direnv: nix-direnv: renewed cache
direnv: export +AR +AS +CC +CONFIG_SHELL +CXX +HOST_PATH +IN_NIX_SHELL +LD +NIX_BINTOOLS +NIX_BINTOOLS_WRAPPER_TARGET_HOST_x86_64_unknown_linux_gnu +NIX_BUILD_CORES +NIX_CC +NIX_CC_WRAPPER_TARGET_HOST_x86_64_unknown_linux_gnu +NIX_CFLAGS_COMPILE +NIX_ENFORCE_NO_NATIVE +NIX_HARDENING_ENABLE +NIX_INDENT_MAKE +NIX_LDFLAGS +NIX_STORE +NM +OBJCOPY +OBJDUMP +RANLIB +READELF +SIZE +SOURCE_DATE_EPOCH +STRINGS +STRIP +buildInputs +buildPhase +builder +configureFlags +depsBuildBuild +depsBuildBuildPropagated +depsBuildTarget +depsBuildTargetPropagated +depsHostHost +depsHostHostPropagated +depsTargetTarget +depsTargetTargetPropagated +doCheck +doInstallCheck +dontAddDisableDepTrack +name +nativeBuildInputs +out +outputs +patches +phases +propagatedBuildInputs +propagatedNativeBuildInputs +shell +shellHook +stdenv +strictDeps +system ~PATH ~XDG_DATA_DIRS
$ # no spurious refresh this time
$ 
```

**Comment by bbenne10** [comment]
I am not sure how I missed this issue. Sorry for that!

I'll look this over and see if I can come to a conclusion soon (maybe this evening).

Thanks for reporting!

**Comment by PaulGrandperrin** [comment]
Hi! What's the status?

I had a few issues with `nix-direnv` messing with my `flake.lock` when doing things with `git rebase -i` and the solution has be to add `--no-write-lock-file` everywhere.

I'm wondering if `--no-write-lock-file` shouldn't be the default?

I feel that many things can go wrong when `flake.lock` can be silently written at every prompt.

**Comment by alexshpilkin** [comment]
I know @PaulGrandperrin has already seen that, but I’ll mention just for reference that with the merge of https://github.com/NixOS/nix/pull/6917 it should now be much easier to keep `flake.lock` out of version control entirely. (Doesn’t solve the `git rebase -i` problem with a committed `flake.lock`, though.)

**Comment by bbenne10** [comment]
I'm sorry - life's gotten rather busy lately. I'd happily accept a PR for this if someone has something locally.

I'm not entirely sure if I know how we should fix this.
I *think* I believe that `--no-write-lock-file` should be default (we should be representing the current state of the lock rather than moving it forward), but I don't want to introduce problems via the new flag (but I will admit that I don't see how it *would*).

I wonder if @Mic92 has any thoughts on this?

**Comment by PaulGrandperrin** [comment]
Ahah, you are on the other issue too 🙂
Yes, I'm just trying to be extra safe, because it really bit me hard a few times!

Yes the fix in `nix` solves 90% of the problem but indeed if the file is already committed, it can still be annoying.

And in the end, i just feel that something that runs at every prompt should never ever touch the files that the user is working on. As a general rule.

**Comment by PaulGrandperrin** [comment]
But if you disagree, that's really fine, because i can manually fix all my `.envrc` and forget about it 😉

**Comment by bbenne10** [comment]
I will say that I think we'll need to prune `$@` as we do in `use_nix`, unless it is safe to proxy ALL arguments through there. I *do* think though that `--no-update-lock-file` might be a better option *and* would be explicitly safe (by disallowing rather than simply not writing any changes to the lock file). Does any one see any issues with simply appending `--no-update-lock-file` to `nix flake archive ...` as linked above?

**Comment by PaulGrandperrin** [comment]
`--no-update-lock-file` doesn't work for me because it means that I'll often lose the devShell env with `error: flake '...' requires lock file changes but they're not allowed due to '--no-update-lock-file'`.

It'll usually make me loose the env when I'm in the middle of doing something, and fixing the `flake.lock` might not be easy at that time, but I still need the env to work. A kind of chicken and egg problem.

I might have specials needs though, I really don't know.
And it's really easy to fix to add `use flake --no-write-lock-file`, so it's not a big deal for me.

**Comment by PaulGrandperrin** [comment]
Maybe a middle ground would be to:
- make `direnv-nix` never ever touch files (to prevent surprises and issues)
- always work when reasonable (so, even with a missing or outdated `flake.lock`)
- but print a warning when the `flake.lock` is missing or out of date

**Comment by bbenne10** [comment]
Okay - I agree that `nix-direnv` should not update the `flake.lock` (though I'm not sure if that's what you mean by "never ever touch files". That would prevent creating the gcroot, for instance.) I will try to have a look over my lunch break today and see if I can rework the approach to work as above, but I can't make many promises. 

I think that for the time being it might be a better solution to simply say that you should append `--no-write-lock-file` to your `.envrc`. This should continue working with any future work we do upstream, so you'll have to do it only once. I have this issue on my radar, but I'm finding free time very hard to come by these days.

**Comment by PaulGrandperrin** [comment]
Yes, that's what I meant by "never touching files". I just said that because I that what I would expect from that kind of "background" software.

To be honest, I'm a kinda recent NixOS user and even more so `nix-direnv` user, so maybe I don't have enough experience with it.

He's my strong use case to explain my why I fear silent changes: I use extremely often `git rebase -i` and that kind of silent changes can really mess things up ;)

But yes, I'm very happy that it's so easy to apply custom arguments to nix ! Great project!
